Forum: initial service

GeoffreyHuck opened the pull request:
fixes #884 

Almost done, a few questions/observations:

- One big limitation of the actual query system is that we cannot make complex "OR" queries like here: https://gorm.io/docs/advanced_query.html#Group-Conditions . It looks like the GORM version we're using is too old and doesn't support it. For this reason, the current code inefficiently uses multiple queries until we'll update it.

After a few checks:
- We need to investigate why we rely on a fork of an old version of gorm (https://github.com/France-ioi/gorm/commits?author=zenovich), and if we still need this change if we update
- Using a recent version of gorm requires us to upgrade go to a later version





- Questions for the restriction number 3:
```
- the following rules all matches:
  * the current-user is descendant of the thread helper_group
  * the thread is either open (=waiting_for_participant or =waiting_for_trainer), or closed for less than 2 weeks
  * the current-user has validated the item
```

1). If the current-user didn't validate the item but is in a team who did validate it, do we get the thread?
2). If there is a thread and it's closed for more than 2 weeks, do we return "Access denied", or status=not_started?
3). Do the 2 weeks rule also apply with the other restrictions which are:
```
- the current-user is the thread participant and allowed to "can_view >= content" the item
- the current-user has the "can_watch >= answer" permission on the item
```




- Questions for the tests:
Some tests are failing because they check the complete status of the table after a deletion

eg.
```
    And the table "groups_ancestors" should be:
      | ancestor_group_id | child_group_id | is_self |
      | 13                | 13             | 1       |
      | 14                | 14             | 1       |
      | 14                | 21             | 0       |
      | 15                | 13             | 0       |
      | 15                | 15             | 1       |
      | 21                | 21             | 1       |
      | 22                | 13             | 0       |
      | 22                | 14             | 0       |
      | 22                | 21             | 0       |
      | 22                | 22             | 1       |
```

 It is verified that the table contains all the rows that were not deleted, and those are copy-pasted. If we add new rows in the tables to check new functionnalities, those tests will always fail and have to be update.

A solution could be to add a new Gherkin feature `And the table "groups_groups" should stay unchanged but the rows with child_group_id "11" should be deleted`. What do you think ?

smadbe commented:
>     One big limitation of the actual query system is that we cannot make complex "OR" queries like here: https://gorm.io/docs/advanced_query.html#Group-Conditions . It looks like the GORM version we're using is too old and doesn't support it. For this reason, the current code inefficiently uses multiple queries until we'll update it.

For info, Dmitry started working on updating gorm but faced several issues, see #769 . 

> We need to investigate why we rely on a fork of an old version of gorm (https://github.com/France-ioi/gorm/commits?author=zenovich), and if we still need this change if we update

The fork was required for advanced features (or bug-fixes?) which were missing... to be checked. I suspect it may be a lot of work.

> Using a recent version of gorm requires us to upgrade go to a later version

In itself, there is no problem in upgrading Go to a later version.

> 1). If the current-user didn't validate the item but is in a team who did validate it, do we get the thread?

I omitted the team aspect on purpose for now for simplifying the problem as team may require to use the service with specific args (apply to most existing services).

> 2). If there is a thread and it's closed for more than 2 weeks, do we return "Access denied", or status=not_started?

denied

> 3). Do the 2 weeks rule also apply with the other restrictions which are:
> - the current-user is the thread participant and allowed to "can_view >= content" the item
> - the current-user has the "can_watch >= answer" permission on the item

no

> A solution could be to add a new Gherkin feature And the table "groups_groups" should stay unchanged but the rows with child_group_id "11" should be deleted. What do you think ?

You can add new Gherkin rules, just keep in mind that the goal of these tests is that they should be readable without re-reading their spec each time. In your case, if there are no strange case I miss, that looks good, it improves both readability and writability.

GeoffreyHuck commented:
I checked the problem with lint. When connecting though ssh to reproduce I faced a "go not found" error, and indeed the path to go was not in PATH. I don't know why it sometimes worked before. It seems to work for now at least.

I added -coverpkg option for go test, which makes the cover go from one package to another to avoid testing the same things multiple times. I then faced a problem with the *test packages not being 100% covered. So in the end I use "go list" to get list of the packages of the project, and remove those ending with "test" from the list.

I also added a Gherkin feature to test for deletion of rows and refactored the database-related Gherkin features because the code was a little messy and it didn't pass the lint.

Finally, I added a pre-commit hook to check the validity of the circleci config if it is modified.

smadbe commented on app/database/thread_store.go:
```diff
@@ -0,0 +1,99 @@
+		Joins("JOIN groups_ancestors_active ON groups_ancestors_active.child_group_id = ?", user.GroupID).
+		Where("threads.helper_group_id = groups_ancestors_active.ancestor_group_id").
+		Where("threads.item_id = ?", itemID).
+		Where("threads.status != 'closed' OR (threads.status = 'closed' AND threads.latest_update_at > ?)", twoWeeksAgo).
```
Probably does not really matter in terms of perf but equivalent to (the simpler):
```suggestion
		Where("threads.status != 'closed' OR threads.latest_update_at > ?", twoWeeksAgo).
```

smadbe commented on app/database/thread_store.go:
```diff
@@ -0,0 +1,99 @@
+		Limit(1).
+		PluckFirst("status", &status).
+		Error()
+	if err != nil {
```
don't you have the risk here to interpret other errors (e.g., db connection error) to a `not_started` status?
An (ugly?) alternative might be to "union" your query with `select 'not_started' as status`.

GeoffreyHuck replied:
fixed. Added check for IsRecordNotFoundError

smadbe commented on db/migrations/2342234112_delete_old_forum_tables.sql:
```diff
@@ -0,0 +1,56 @@
+-- +migrate Up
```
The migration file name were corresponding to YYMMDDhhmm ... I suppose "42" is not the month, what is the logic?

GeoffreyHuck replied:
The logic was random ^^'. I edited the names and will change the few others I made in previous changes for consistency.

smadbe reviewed (commented):
Great! Look good 👍 
Just a few remarks/questions.

GeoffreyHuck commented on app/api/items/get_thread.robustness.feature:
```diff
@@ -0,0 +1,152 @@
+      Should be forbidden when
+      the current-user is descendant of the thread helper group
+      and user has validated the item
+      but the thread is closed for 2 weeks
```
Yes, it's two weeks exactly (==). And the next one is to test that it works for more (>). I also wondered if we really need to test this.

smadbe reviewed (approved):
Ok, good for me! 

(just https://github.com/France-ioi/AlgoreaBackend/pull/893#discussion_r1121429392 maybe)
